src/circuit/circuit/stream-handler.js

```diff
@@ -1,139 +1,88 @@
   /**
    * Create a stream handler for connection
    *
-   * @param {Connection} conn - connection to read/write
-   * @param {Function|undefined} cb - handshake callback called on error
-   * @param {Number} timeout - handshake timeout
-   * @param {Number} maxLength - max bytes length of message
+   * @param {object} options
+   * @param {*} options.stream - A duplex iterable
+   * @param {Number} options.maxLength - max bytes length of message
    */
-  constructor (conn, cb, timeout, maxLength) {
-    this.conn = conn
-    this.stream = null
-    this.shake = null
-    this.timeout = cb || 1000 * 60
-    this.maxLength = maxLength || 4096
+  constructor ({ stream, maxLength = 4096 }) {
+    this.stream = stream
 
-    if (typeof cb === 'function') {
-      this.timeout = timeout || 1000 * 60
-    }
-
-    this.stream = handshake({ timeout: this.timeout }, cb)
-    this.shake = this.stream.handshake
-
-    pull(this.stream, conn, this.stream)
-  }
-
-  isValid () {
-    return this.conn && this.shake && this.stream
+    this.shake = handshake(this.stream)
+    this.decoder = lp.decode.fromReader(this.shake.reader, { maxDataLength: maxLength })
   }
 
   /**
    * Read and decode message
-   *
-   * @param {Function} cb
-   * @returns {void|Function}
+   * @async
+   * @returns {void}
    */
-  read (cb) {
-    if (!this.isValid()) {
-      return cb(new Error('handler is not in a valid state'))
+  async read () {
+    const msg = await this.decoder.next()
+    if (msg.value) {
+      const value = CircuitPB.decode(msg.value.slice())
+      log('read message type', value.type)
+      return value
     }
 
-    lp.decodeFromReader(
-      this.shake,
-      { maxLength: this.maxLength },
-      (err, msg) => {
-        if (err) {
-          log.err(err)
-          // this.shake.abort(err)
-          return cb(err)
-        }
-
-        return cb(null, msg)
-      })
+    log('read received no value, closing stream')
+    // End the stream, we didn't get data
+    this.close()
   }
 
   /**
    * Encode and write array of buffers
    *
-   * @param {Buffer[]} msg
-   * @param {Function} [cb]
-   * @returns {Function}
+   * @param {*} msg An unencoded CircuitRelay protobuf message
    */
-  write (msg, cb) {
-    cb = cb || (() => {})
-
-    if (!this.isValid()) {
-      return cb(new Error('handler is not in a valid state'))
-    }
-
-    pull(
-      values([msg]),
-      lp.encode(),
-      collect((err, encoded) => {
-        if (err) {
-          log.err(err)
-          this.shake.abort(err)
-          return cb(err)
-        }
-
-        encoded.forEach((e) => this.shake.write(e))
-        cb()
-      })
-    )
+  write (msg) {
+    log('write message type %s', msg.type)
+    this.shake.write(lp.encode.single(CircuitPB.encode(msg)))
   }
 
   /**
    * Get the raw Connection
    *
-   * @returns {null|Connection|*}
+   * @returns {*} A duplex iterable
    */
   getRawConn () {
-    return this.conn
+    return this.stream
```

Comment:
Can we leverage this to better name this function?

Reply:
It's not being used anywhere so I will just remove it.